cmd/serverless-init/log/channel_writer.go

```diff
@@ -0,0 +1,74 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package log
+
+import (
+	"bytes"
+	"io"
+
+	logConfig "github.com/DataDog/datadog-agent/comp/logs/agent/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ChannelWriter is a buffered writer that log lines (lines ending with a \n) to a channel
+// to be sent to our intake.
+type ChannelWriter struct {
+	Buffer  bytes.Buffer
+	Channel chan *logConfig.ChannelMessage
+	IsError bool
+}
+
+// NewChannelWriter returns a new channel writer.
+// Implements io.Writer, used for redirecting stdout/stderr
+// logs to Datadog.
+func NewChannelWriter(ch chan *logConfig.ChannelMessage, isError bool) *ChannelWriter {
+	return &ChannelWriter{
+		Channel: ch,
+		IsError: isError,
+	}
+}
+
+// Write buffers Writes from our stdout/stderr fd,
+// and sends to the channel once we've received newlines.
+func (cw *ChannelWriter) Write(p []byte) (n int, err error) {
+	n, err = cw.Buffer.Write(p)
```

Comment:
From the Buffer.Write docs
```
// Write appends the contents of p to the buffer, growing the buffer as
// needed. The return value n is the length of p; err is always nil. If the
// buffer becomes too large, Write will panic with ErrTooLarge.
```

```suggestion
	n, _ = cw.Buffer.Write(p)
```

it could be fun to add a test to replicate the buffer panic, in case we want to rescue it. But probably not that important

Reply:
Yeah I'm guessing that's the case where we the malloc fails due to memory pressure. I agree that I don't think it's probably the most important.